File: MinimumDistanceBetweenBSTNodes.cpp
/* 
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int ans=100000;
public:
    int minDiffInBST(TreeNode* root) {
        dfs(root);
        return ans;
    }
    
    TreeNode* rightChild(TreeNode*node){
        return (node->right ? rightChild(node->right) : node);
    }
    TreeNode* leftChild(TreeNode*node){
        return (node->left ? leftChild(node->left) : node);
    }
    
    void dfs(TreeNode* root){
        if(!root)
            return ;
        if(root->left && root->right){
            ans=min(ans, root->val - rightChild(root->left)->val);
            ans=min(ans, leftChild(root->right)->val - root->val);
            
            dfs(root->left);
            dfs(root->right);
        }else if(root->left){
            ans=min(ans, root->val - rightChild(root->left)->val);
            
            dfs(root->left);
        }else if(root->right){
            ans=min(ans, leftChild(root->right)->val - root->val);
            
            dfs(root->right);
        }else // leaf node
            return;
    }
};

File: MinimumDistanceBetweenBSTNodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        order = []
        def dfs(node):
            if node:
                dfs(node.left)
                order.append(node.val)
                dfs(node.right)


        dfs(root)

        diff = 999999
        for i in range(len(order)-1):
            diff = min(diff, order[i+1]-order[i])

        return diff
